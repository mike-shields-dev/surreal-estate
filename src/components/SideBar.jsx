import React, { useState } from "react";
import PropTypes from "prop-types";
import qs from "qs";
import { Link, useLocation, useNavigate } from "react-router-dom";
import { FaPlus, FaSearch } from "react-icons/fa";
import css from "../styles/SideBar.module.css";
import TextBox from "./TextBox";
import useToggle from "../hooks/useToggle";

const SideBar = ({ cities }) => {
  const { status: isExpanded, toggleStatus: toggleExpanded } = useToggle(false);
  const navigate = useNavigate();
  const { search } = useLocation();
  const [titleSearch, setTitleSearch] = useState("");
  const currentUrlParams = qs.parse(search, {
    ignoreQueryPrefix: true,
  });
  const { query, sort } = currentUrlParams;

  const buildParamsString = (operation, valueObj) => {
    const newUrlParams = {
      ...currentUrlParams,
      [operation]: JSON.stringify({
        ...JSON.parse(currentUrlParams[operation] || "{}"),
        ...valueObj,
      }),
    };
    return qs.stringify(newUrlParams, {
      addQueryPrefix: true,
      encode: false,
    });
  };

  const handleSearchChange = (event) => {
    setTitleSearch(event.target.value);
  };

  const handleSearchSubmit = (event) => {
    event.preventDefault();
    let newUrlParamsString;
    if (titleSearch) {
      newUrlParamsString = buildParamsString("query", {
        title: { $regex: titleSearch },
      });
    } else {
      newUrlParamsString = buildParamsString("query", {
        title: { $regex: ".*" },
      });
    }
    navigate(newUrlParamsString);
  };

  const handleReset = () => {
    navigate("");
    setTitleSearch("");
  };

  return (
    <aside
      className={`
        ${css.sidebar} 
        ${css[`sidebar${isExpanded ? "--open" : "--closed"}`]}`}
    >
      <button
        onClick={toggleExpanded}
        type="button"
        className={css["sidebar__toggle-button"]}
      >
        <span className="visibly-hidden">Menu</span>
        <FaPlus
          aria-hidden="true"
          className={`
            ${css.sidebar__icon}
            ${css[`sidebar__icon${isExpanded ? "--open" : "--closed"}`]}`}
          title={isExpanded ? "close menu" : "open menu"}
        />
      </button>

      <section className={css.sidebar__section}>
        <form className={css.sidebar__form} onSubmit={handleSearchSubmit}>
          <TextBox
            field="title"
            onChange={handleSearchChange}
            value={titleSearch}
            type="search"
          />
          <button type="submit">
            <FaSearch />
            <span className={css["sidebar__search-submit-button"]}>Search</span>
          </button>
        </form>
      </section>
      <section className={css.sidebar__section}>
        <h3 className={css.sidebar__heading}>City</h3>
        <ul className={css["sidebar__link-list"]}>
          {cities.map((city) => {
            const queryString = buildParamsString("query", { city });
            return (
              <li key={`${city}-city-filter-link`}>
                <Link
                  className={
                    css[
                      `sidebar__link${
                        query && query.includes(`"city":"${city}"`)
                          ? "--active"
                          : ""
                      }`
                    ]
                  }
                  to={queryString}
                >
                  {city}
                </Link>
              </li>
            );
          })}
        </ul>
      </section>
      <section className={css.sidebar__section}>
        <h3 className={css.sidebar__heading}>Price</h3>
        <ul className={css["sidebar__link-list"]}>
          <li>
            <Link
              className={
                css[
                  `sidebar__link${
                    sort && sort.includes(`"price":1`) ? "--active" : ""
                  }`
                ]
              }
              to={buildParamsString("sort", { price: 1 })}
            >
              Ascending
            </Link>
          </li>
          <li>
            <Link
              className={
                css[
                  `sidebar__link${
                    sort && sort.includes(`"price":-1`) ? "--active" : ""
                  }`
                ]
              }
              to={buildParamsString("sort", { price: -1 })}
            >
              Descending
            </Link>
          </li>
        </ul>
      </section>
      <button
        className={css["sidebar__reset-button"]}
        type="button"
        onClick={handleReset}
      >
        reset
      </button>
    </aside>
  );
};

SideBar.propTypes = {
  cities: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default SideBar;
