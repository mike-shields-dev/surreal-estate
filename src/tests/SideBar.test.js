import React from "react";
import { BrowserRouter } from "react-router-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import SideBar from "../components/SideBar";

const cities = ["Leeds", "Liverpool", "Manchester", "Sheffield"];
const props = {
  cities,
};

const renderSideBar = () =>
  render(
    <BrowserRouter>
      <SideBar {...props} />
    </BrowserRouter>
  );

describe("SideBar", () => {
  it("matches snapshot", () => {
    const { asFragment } = renderSideBar();

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders menu button", () => {
    renderSideBar();
    const menuButton = screen.getByRole("button", { name: /menu/i });

    expect(menuButton).toBeInTheDocument();
  });

  it("menu button opens and closes the sidebar", async () => {
    renderSideBar();
    const menuButtonEl = screen.getByRole("button", { name: /menu/i });
    const sideBarEl = await screen.findByRole("complementary");

    expect(sideBarEl.getAttribute("class")).toMatch(/closed/i);

    fireEvent.click(menuButtonEl);

    expect(sideBarEl.getAttribute("class")).toMatch(/open/i);

    fireEvent.click(menuButtonEl);

    expect(sideBarEl.getAttribute("class")).toMatch(/closed/i);
  });

  it("renders a heading of city for the city links", () => {
    renderSideBar();

    expect(screen.getByRole("heading", { name: /city/i })).toBeInTheDocument();
  });

  it("renders a link for each of the given cities", () => {
    renderSideBar();

    const cityLinks = screen
      .getAllByRole("link")
      .filter((cityLink) => !cityLink.hasAttribute("href", props.pathname));

    cityLinks.forEach((cityLink, i) => {
      expect(cities.includes(cityLink.value)).toBeTruthy();
      expect(cityLink).toHaveAttribute(
        "href",
        `${props.pathname}${props.cities[i]}`
      );
    });
  });

  it("applies active class ONLY to the city link that was clicked", () => {
    renderSideBar();

    const cityLinks = screen
      .getAllByRole("link")
      .filter((link) => link.href.match(/city/i));

    cityLinks.forEach(async (cityLink) => {
      fireEvent.click(cityLink);

      expect(cityLink.getAttribute("class")).toMatch(/active/i);

      const inactiveCityLinks = cityLinks.filter(
        (link) => !link.className.match(/active/i)
      );

      expect(inactiveCityLinks).toHaveLength(cityLinks.length - 1);
      expect(inactiveCityLinks.includes(cityLink)).toBeFalsy();
    });
  });

  it("renders a heading of pric for the price sorting links", () => {
    renderSideBar();

    expect(screen.getByRole("heading", { name: /price/i })).toBeInTheDocument();
  });

  it("renders links for sorting by price", () => {
    renderSideBar();

    expect(
      screen.getByRole("link", { name: /ascending/i })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("link", { name: /descending/i })
    ).toBeInTheDocument();
  });

  it("applies active class ONLY to the city link that was clicked", async () => {
    renderSideBar();

    const sortPriceLinkAsc = screen.getByRole("link", { name: /ascending/i });
    const sortPriceLinkDesc = screen.getByRole("link", { name: /descending/i });

    fireEvent.click(sortPriceLinkAsc);

    expect(sortPriceLinkAsc.getAttribute("class")).toMatch(/active/i);
    expect(sortPriceLinkDesc.getAttribute("class")).not.toMatch(/active/i);

    fireEvent.click(sortPriceLinkDesc);

    expect(sortPriceLinkDesc.getAttribute("class")).toMatch(/active/i);
    expect(sortPriceLinkAsc.getAttribute("class")).not.toMatch(/active/i);
  });
});
