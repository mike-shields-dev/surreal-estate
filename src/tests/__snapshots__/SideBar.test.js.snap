// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar matches snapshot 1`] = `
<DocumentFragment>
  <aside
    class="
        sidebar 
        sidebar--closed"
  >
    <button
      class="sidebar__button"
      type="button"
    >
      <span
        class="visibly-hidden"
      >
        Menu
      </span>
      <svg
        aria-hidden="true"
        class="
            sidebar__icon
            sidebar__icon--closed"
        fill="currentColor"
        focusable="false"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          open menu
        </title>
        <path
          d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
        />
      </svg>
    </button>
    <section
      class="sidebar__section"
    >
      <form
        class="sidebar__form"
      >
        <div
          class="textbox"
        >
          <label
            class="textbox__label"
            for="title"
          >
            title
          </label>
          <input
            id="title"
            name="title"
            type="search"
            value=""
          />
        </div>
        <button
          type="submit"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
          <span
            class="sidebar__search-submit-button"
          >
            Search
          </span>
        </button>
      </form>
    </section>
    <br />
    <hr
      class="sidebar__separator"
    />
    <br />
    <section
      class="sidebar__section"
    >
      <h3
        class="sidebar__heading"
      >
        City
      </h3>
      <ul
        class="sidebar__link-list"
      >
        <li>
          <a
            class="sidebar__link"
            href="/?query={\\"city\\":\\"Leeds\\"}"
          >
            Leeds
          </a>
        </li>
        <li>
          <a
            class="sidebar__link"
            href="/?query={\\"city\\":\\"Liverpool\\"}"
          >
            Liverpool
          </a>
        </li>
        <li>
          <a
            class="sidebar__link"
            href="/?query={\\"city\\":\\"Manchester\\"}"
          >
            Manchester
          </a>
        </li>
        <li>
          <a
            class="sidebar__link"
            href="/?query={\\"city\\":\\"Sheffield\\"}"
          >
            Sheffield
          </a>
        </li>
      </ul>
    </section>
    <hr
      class="sidebar__separator"
    />
    <br />
    <section
      class="sidebar__section"
    >
      <h3
        class="sidebar__heading"
      >
        Price
      </h3>
      <ul
        class="sidebar__link-list"
      >
        <li>
          <a
            class="sidebar__link"
            href="/?sort={\\"price\\":1}"
          >
            Ascending
          </a>
        </li>
        <li>
          <a
            class="sidebar__link"
            href="/?sort={\\"price\\":-1}"
          >
            Descending
          </a>
        </li>
      </ul>
    </section>
    <hr
      class="sidebar__separator"
    />
    <br />
  </aside>
</DocumentFragment>
`;
